diff --git a/modules/mach/int.mlw b/modules/mach/int.mlw
index 5a3374e..c7de56d 100644
--- a/modules/mach/int.mlw
+++ b/modules/mach/int.mlw
@@ -13,7 +13,8 @@ module Int
   use export int.ComputerDivision
 
   let (/) (x: int) (y: int)
-    requires { y <> 0 } ensures { result = div x y }
+    requires { "expl:division by zero" y <> 0 } 
+    ensures  { result = div x y }
   = div x y
 
 end
diff --git a/theories/map.why b/theories/map.why
index c9c9cca..235b944 100644
--- a/theories/map.why
+++ b/theories/map.why
@@ -219,7 +219,7 @@ theory Occ
   lemma occ_bounds:
     forall v: 'a, m: map int 'a, l u: int.
     l <= u -> 0 <= occ v m l u <= u - l
-    (* direct when l>=u, by induction on b when l <= u *)
+    (* direct when l>=u, by induction on u when l <= u *)
 
   lemma occ_append:
     forall v: 'a, m: map int 'a, l mid u: int.
